js_modules/dagit/packages/ui/src/components/TagSelector.tsx

```diff
@@ -0,0 +1,170 @@
+          <Icon name="close" />
+        </Box>
+      </Box>
+    </Tag>
+  );
+};
+
+const defaultRenderDropdownItem = (tag: string, dropdownItemProps: DropdownItemProps) => {
+  return (
+    <label>
+      <MenuItem
+        text={
+          <Box flex={{alignItems: 'center', gap: 8}}>
+            <Checkbox checked={dropdownItemProps.selected} onChange={dropdownItemProps.toggle} />
+            <span>{tag}</span>
+          </Box>
+        }
+      />
+    </label>
+  );
+};
+
+export const TagSelector = ({
+  allTags,
+  placeholder,
+  selectedTags,
+  setSelectedTags,
+  renderTag,
+  renderDropdownItem,
+  renderDropdown,
+  dropdownStyles,
+  renderTagList,
+}: Props) => {
+  const [isDropdownOpen, setIsDropdownOpen] = React.useState(false);
+  const dropdown = React.useMemo(() => {
+    const dropdownContent = (
+      <Box
+        style={{
+          maxHeight: '500px',
+          overflowY: 'auto',
+          ...dropdownStyles,
+        }}
+      >
+        {allTags.map((tag) => {
+          const selected = selectedTags.includes(tag);
+          const toggle = () => {
+            setSelectedTags(
+              selected ? selectedTags.filter((t) => t !== tag) : [...selectedTags, tag],
+            );
+          };
+          if (renderDropdownItem) {
+            return <div key={tag}>{renderDropdownItem(tag, {toggle, selected})}</div>;
+          }
+          return defaultRenderDropdownItem(tag, {toggle, selected});
+        })}
+      </Box>
+    );
+    if (renderDropdown) {
+      return renderDropdown(dropdownContent);
+    }
+    return <Menu>{dropdownContent}</Menu>;
+  }, [allTags, dropdownStyles, renderDropdown, renderDropdownItem, selectedTags, setSelectedTags]);
+
+  const dropdownContainer = React.useRef<HTMLDivElement>(null);
+
+  const tagsContent = React.useMemo(() => {
+    if (selectedTags.length === 0) {
+      return <Placeholder>{placeholder || 'Select tags'}</Placeholder>;
+    }
+    const tags = selectedTags.map((tag) =>
+      (renderTag || defaultRenderTag)(tag, {
+        close: (ev) => {
+          setSelectedTags((tags) => tags.filter((t) => t !== tag));
+          ev.stopPropagation();
+        },
+      }),
+    );
+    if (renderTagList) {
+      return renderTagList(tags);
+    }
+    return tags;
+  }, [placeholder, selectedTags, renderTag, renderTagList]);
+
+  return (
+    <Popover
+      placement="bottom"
+      isOpen={isDropdownOpen}
+      onInteraction={(nextOpenState, e) => {
+        const target = e?.target;
+        if (isDropdownOpen && target instanceof HTMLElement) {
+          const isClickInside = dropdownContainer.current?.contains(target);
+          if (!isClickInside) {
+            setIsDropdownOpen(nextOpenState);
+          }
+        }
+      }}
+      content={<div ref={dropdownContainer}>{dropdown}</div>}
+      targetTagName="div"
+    >
+      <Box
+        as={Container}
+        padding={{vertical: 4, horizontal: 6 as any}}
```

Comment:
I'm always skeptical of breaking out of 4px-based values...maybe just use a `div` for this instead of a `Box`?